tealer/analyses/utils/stack_emulator.py

```diff
@@ -0,0 +1,451 @@
+                position = 0,
+            ),
+            KnownStackValue(
+                ins = Mulw(),
+                args = [Int(10), Int(11)],
+                position = 1,
+            )
+
+        Value which is pushed first will have position 0.
+    """
+
+    def __init__(self, ins: Instruction, args: List, position: int = 0) -> None:
+        self._ins = ins
+        self._args: List[Union[KnownStackValue, UnknownStackValue]] = args
+        self._position = position
+
+    @property
+    def instruction(self) -> Instruction:
+        return self._ins
+
+    @property
+    def args(self) -> List[Union["KnownStackValue", UnknownStackValue]]:
+        """List of stack values consumed by self.instruction.
+
+        self.args[0] is the last popped(first pushed) value from the stack.
+        self.args[-1] is the first popped value from the stack:
+            self.args[-1] is on top of the stack right before the execution
+            of self.instruction.
+        """
+        return self._args
+
+    @property
+    def position(self) -> int:
+        return self._position
+
+    def __str__(self) -> str:
+        # args = ", ".join(str(i) for i in self._args)
+        # return f"{self._ins.__class__.__name__}(<{args}>)"
+        args = "[" + ", ".join(str(i) for i in self.args) + "]"
+        # return f"KnownStackValue(ins = {str(self._ins.__class__.__name__)}, args = {args})"
+        return f"KnownStackValue(ins = {repr(self._ins)}, args = {args})"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+StackValue = Union[KnownStackValue, UnknownStackValue]
+
+
+class Stack:
+    """Represent a stack.
+
+    Elements are stored in a list. Top of the stack is at the end of the List.
+
+    Stack is assumed to have infinite UnknownStackValue below the current pointer.
+    """
+
+    def __init__(self) -> None:
+        self._values: List[StackValue] = []
+
+    def push(self, value: StackValue) -> None:
+        self._values.append(value)
+
+    def pop(self, count: int) -> List[StackValue]:
+        """Pop elements from the stack.
+
+        Args:
+            count: Number of elements to pop.
+
+        Returns:
+            list of popped values. First popped value is at the end of the list.
+        """
+        if count == 0:
+            return []
+        if len(self._values) >= count:
+            vals = self._values[-count:]
+            self._values = self._values[:-count]
+            return vals
+        remaining_elements: List[StackValue] = [
+            UnknownStackValue() for _ in range(count - len(self._values))
+        ]
+        vals = remaining_elements + self._values
+        self._values = []
+        return vals
+
+    def __str__(self) -> str:
+        return "[" + ",".join([str(value) for value in self._values]) + "]"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+@lru_cache(maxsize=None)
+def emulate_stack(bb: BasicBlock) -> Dict[Instruction, KnownStackValue]:
```

Comment:
I think we should rename `stack_emulator` and this function to something more explicit. Maybe `expression_extractor` / `stack_ast_builder`/  or something similar? (abstract value tree?)